[147]: Added logic to store the block headers

Natsvora opened the pull request:
[147]: Added logic to store the block headers

Commit:
[147]: Added logic to store the block headers

dambaev commented on backend/src/api/blocks.ts:
```diff
@@ -25,8 +25,9 @@ import DifficultyAdjustmentsRepository from '../repositories/DifficultyAdjustmen
+import opBlockHeaderService from '../oe/service/op-block-header.service';
 
-class Blocks {
+class Blocks {  
```
```
git diff --check
```
is able to highlight the trailing spaces that needed to be removed

Natsvora replied:
Fixed

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,91 @@
+        await opBlockHeaderRepository.$getLatestBlockHeight();
+      const currentBlockHeight =
+        (await bitcoinApi.$getBlockHeightTip()) -
+        config.MEMPOOL.INITIAL_BLOCKS_AMOUNT;
```
MEMPOOL.INITIAL_BLOCKS_AMOUNT config variable defines how many blocks should be sent to websocket client at the start of the websocket connection. and it's default value is 8 ( https://github.com/dambaev/op-energy/blob/a8bbda9d2629034731830108de62450b7de9ac96/backend/src/config.ts#L136 )

I assume, that you want to use this config var to implement 6-blocks rule, in this case I suggest to introduce separate variable with value 6, like OP_ENERGY_BLOCK_CONFIRMATIONS_AMOUNT

Natsvora replied:
Added a variable

dambaev commented on backend/src/api/blocks.ts:
```diff
@@ -462,6 +463,9 @@ class Blocks {
           await blocksRepository.$saveBlockInDatabase(blockExtended);
 
+          // Currently Commented it as during every main loop update block headers are already syncing
+          // await opBlockHeaderService.$saveBlockHeader(this.currentBlockHeight);
```
1. pushing commented code is not a good idea: if it shouldn't be running, then it shouldn't be here, as for me;
2. opBlockHeaderService.$saveBlockHeader() is not total, meaning, that it will only persist `argument - 6` block: 
$syncOlderBlockHeader() is able to persist [last_persisted_block;current_tip - 8] which can handle case of, say restarting of bitcoin node after hour of being offline or due to out of space event... so $syncOlderBlockHeader() is a better candidate as an entry point to persistence layer

Natsvora replied:
Right, syncOlderBlockHeader is the only function which will be expose outside the class and at persistence layer

dambaev commented on backend/src/api/blocks.ts:
```diff
@@ -25,8 +25,9 @@ import DifficultyAdjustmentsRepository from '../repositories/DifficultyAdjustmen
 import PricesRepository from '../repositories/PricesRepository';
 import priceUpdater from '../tasks/price-updater';
 import oeBlocks from './../oe/api/blocks';
+import opBlockHeaderService from '../oe/service/op-block-header.service';
```
should be removed as well

Natsvora replied:
Done

dambaev commented on backend/src/index.ts:
```diff
@@ -176,6 +177,7 @@ class Server {
 
       await opEnergyIndex.runMainUpdateLoop();
 
+      await opBlockHeaderService.$syncOlderBlockHeader();
```
1. opEnergyIndex has runMainUpdateLoop() already: so this line should be moved there;
2. but I think, that it is better to call $syncOlderBlockHeader() from 2 functions:
1. setUpHttpApiRoutes at https://github.com/dambaev/op-energy/blob/a8bbda9d2629034731830108de62450b7de9ac96/backend/src/oe/index.ts#L14 which will be called before entering the main loop
2. handleNewBlock at https://github.com/dambaev/op-energy/blob/a8bbda9d2629034731830108de62450b7de9ac96/backend/src/oe/index.ts#L32 which will be called each time new block will be discovered

this way bitcoin node will not be unnecessary questioned with current tip request as well as DB query....
even more: handleNewBlock() is already containing block header of discovered block, which means, that it can be used as current tip instead of performing new api call to bitcoin node

Natsvora replied:
Make sense updated the code

dambaev commented:
Case 1
so let's evaluate https://github.com/dambaev/op-energy/pull/271/commits/a8bbda9d2629034731830108de62450b7de9ac96#diff-3e825b8e22c091eb61118b8a63490f949e511593ed85713ad74951b7f39f4a35R51 while assuming:
1. opBlockHeaderRepository.$getLatestBlockHeight() returning `10`;
2. bitcoinApi.$getBlockHeightTip() returns '17'

The expected behavior for those inputs is to assume, that block 11 is now confirmed by existing 6 newer blocks, so block 11 is expected to be persisted.

this gives us currentBlockHeight at https://github.com/dambaev/op-energy/pull/271/commits/a8bbda9d2629034731830108de62450b7de9ac96#diff-3e825b8e22c091eb61118b8a63490f949e511593ed85713ad74951b7f39f4a35R57 will have value 17 - 8 = 9 and while loop's condition will not be satisfied at https://github.com/dambaev/op-energy/pull/271/commits/a8bbda9d2629034731830108de62450b7de9ac96#diff-3e825b8e22c091eb61118b8a63490f949e511593ed85713ad74951b7f39f4a35R76

for me, it seems, like incorrect behavior

case 2
assume:
1. opBlockHeaderRepository.$getLatestBlockHeight() returning `10`;
2. bitcoinApi.$getBlockHeightTip() returns '19'

which gives us currentBlockHeight = 11 and while loop condition will be met exactly 1 time
https://github.com/dambaev/op-energy/pull/271/commits/a8bbda9d2629034731830108de62450b7de9ac96#diff-3e825b8e22c091eb61118b8a63490f949e511593ed85713ad74951b7f39f4a35R76

but then, at https://github.com/dambaev/op-energy/pull/271/commits/a8bbda9d2629034731830108de62450b7de9ac96#diff-3e825b8e22c091eb61118b8a63490f949e511593ed85713ad74951b7f39f4a35R77-R80 you are creating an array of 10 elements, which will make 10 concurrent calls of $saveBlockHeader with arguments [ 10; 20], which makes no sense for me.

I can guess, that 80%-90% of time currentBlockHeight - currentSyncedBlockHeight will be 1, so I assume, that 80%-90% of time main loop will spawn 10 concurrent calls and only 1 of them will be succeeded.

If your idea is to remove $syncOlderBlockHeader() later, then:
1. $saveBlockHeader() should be able to handle case of bitcoin being offline for some time (like out of disk space event), which means, that current tip - currentSyncedBlockHeight can be more than 1
2. I can't agree with the idea of deleting code later: we can deploy new fresh nodes in the future, which means, that some behavior will be different for new instances, which is not expected.

So to sumup: I think, that there should be 1 function instead of $syncOlderBlockHeader() and $saveBlockHeader() (as $saveBlockHeader() just performs check and calls another function). And this function should be called from OpEnergyIndex as I mentioned in the diff commit view

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,91 @@
+
+      while (currentBlockHeight >= currentSyncedBlockHeight) {
+        await Bluebird.map(
+          [...Array(10).keys()].map((i) => i + currentSyncedBlockHeight),
```
```
[...Array( max(1, max(10, currentBlockHeight - currentSyncedBlockHeight ))).keys()].map((i) => i + currentSyncedBlockHeight)
```
?

dambaev replied:
actually:
```
max(1, min(10, currentBlockHeight - currentSyncedBlockHeight ))
```
which defines concurrent jobs of count in  range [1;10]

Natsvora replied:
Done

Natsvora commented:
> Owner

Agree with your comment. I added the suggested changes. That was the bug due to INITIAL_BLOCKS_AMOUNT I fixed it thanks for bringing it out

Commit:
147 Fixed nbdr calculation issue

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,133 @@
+
+      // Only storing 6th block from current tip
+      const blockHash = await bitcoinApi.$getBlockHash(
+        blockHeight - config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT
```
I don't understand why do we decrease blockHeight by config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT here?
It seems, to me, that this function does not what it is being asked for.
What I mean is that as a user, I expect to call `$getBlockHeaderData(1000)` to get block header for block height 1000, not 994.

As for me: caller is responsible for correct arguments to pass, so if caller wants to get block header for block height 1000 it expects exactly this. 
I think, that it will be better to delegate any barriers for opBlockHeaderService.$syncOlderBlockHeader() - this is the place which should keep checks for not calling $saveBlockHeader with arguments [ current_tip - config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT; current_tip].
I would suggest to introduce new type called ConfirmedBlockHeight and a function 
```
verifyConfirmedBlockHeight( blockHeight: number, currentTip: BlockHeight): ConfirmedBlockHeight {
  if( blockHeight <= config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT) {
    throw new Error('block height haven't been confirmed');
  }
  if( blockHeight > currentTip.value - config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT) {
    throw new Error('block height haven't been confirmed');
  }
  return { value: blockHeight };
}
```
thus, you can use ConfirmedBlockHeight as a type for argument to getBlock/saveBlockHeader and co. Maybe everywhere as we suggested, that we only respect confirmed blocks. Anyway, by introducing such type, you will force the caller to construct this value by using function verifyConfirmedBlockHeight() and thus, you can ensure, that caller is giving checked argument.

Natsvora replied:
Good suggestion, Done added

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,133 @@
+      // Adding confirmed block in synced block to get actual synced blocked height
+      let currentSyncedBlockHeight =
+        (await opBlockHeaderRepository.$getLatestBlockHeight()) +
+        config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT;
```
I think it is wrong to add `config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT` here.
```
let currentSyncedBlockHeight = await opBlockHeaderRepository.$getLatestBlockHeight();
```
is the correct one, even more, `opBlockHeaderRepository.$getLatestBlockHeight()` should return value of type `ConfirmedBlockHeight`, because if it had been stored in DB, it had already been confirmed by the caller of the saveBlockHeader() previously.... (that's to refer my suggestion of using ConfirmedBlockHeight)

Natsvora replied:
Done

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,133 @@
+
+      // Only using fetching current block tip if not present in argument
+      if (!currentBlockHeight) {
+        currentBlockHeight = await bitcoinApi.$getBlockHeightTip();
```
```
if (!currentTip) {
  currentTip = await bitcoinApi.$getBlockHeightTip();
}
currentBlockHeight = verifyConfirmedBlockHeight( currentTip  - config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT, currentTip);

```
I would suggest to rename currentBlockHeight into `latestConfirmedBlockHeight` to be more meaningful.

This way, this will be the only place where you are using config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT. I see no reason to use `config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT` in any other places

Natsvora replied:
Done

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,133 @@
+        `Syncing block header from #${currentSyncedBlockHeight} to #${currentBlockHeight}`
+      );
+
+      while (currentBlockHeight >= currentSyncedBlockHeight) {
```
while (latestConfirmedBlockHeight >= currentSyncedBlockHeight)

Natsvora replied:
Done

dambaev commented on backend/src/oe/repositories/OpBlockHeaderRepository.ts:
```diff
@@ -0,0 +1,85 @@
+    blockHeader: BlockHeader
+  ): Promise<void> {
+    
+    const UUID = await opEnergyApiService.$generateRandomHash();
```
UUID is supposed to be used for tracing calls from the entrypoint: boot/mainloop/newBlockHandler/frontend request

so UUID should be the argument of function, it is not supposed to be generated here

Natsvora replied:
Fixed

dambaev commented on backend/src/oe/repositories/OpBlockHeaderRepository.ts:
```diff
@@ -0,0 +1,85 @@
+  public async $getLatestBlockHeight(): Promise<number> {
+    try {
+      const query = `select height from blockheaders order by height desc limit 1`;
+      const UUID = await opEnergyApiService.$generateRandomHash();
```
UUID is supposed to be used for tracing calls from the entrypoint: boot/mainloop/newBlockHandler/frontend request

so UUID should be the argument of function, it is not supposed to be generated here

Natsvora replied:
Fixed

dambaev commented on backend/src/oe/repositories/OpBlockHeaderRepository.ts:
```diff
@@ -0,0 +1,85 @@
+      const params: (number)[] = [
+        height
+      ];
+      const UUID = await opEnergyApiService.$generateRandomHash();
```
UUID is supposed to be used for tracing calls from the entrypoint: boot/mainloop/newBlockHandler/frontend request

so UUID should be the argument of function, it is not supposed to be generated here

Natsvora replied:
Fixed

dambaev commented on backend/src/oe/index.ts:
```diff
@@ -8,12 +8,14 @@ import blocks from '../api/blocks';
   public async setUpHttpApiRoutes( app: Application) {
     opEnergyRoutes.setUpHttpApiRoutes( app);
     await opEnergyApiService.$persistOutcome( "init" );
+    await opBlockHeaderService.$syncOlderBlockHeader();
```
this is the place, where UUID should be sourced from, so this line should be:
```
await opBlockHeaderService.$syncOlderBlockHeader( "init");
```

Natsvora replied:
Fixed

dambaev commented on backend/src/oe/index.ts:
```diff
@@ -31,6 +33,7 @@ class OpEnergyIndex {
 
   async handleNewBlock( block: BlockExtended, txIds: string[], transactions: TransactionExtended[]) {
       await opEnergyApiService.$persistOutcome( "handleNewBlock callback");
+      await opBlockHeaderService.$syncOlderBlockHeader(block.height);
```
this is the place, where UUID should be sourced from, so this line should be:
```
await opBlockHeaderService.$syncOlderBlockHeader("handleNewBlock callback", block.height);
```

Natsvora replied:
Fixed

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,133 @@
+    }
+  }
+
+  public async $syncOlderBlockHeader(currentTip?: number): Promise<void> {
```
should have UUID argument

Natsvora replied:
Added

dambaev commented on backend/src/index.ts:
```diff
@@ -40,6 +40,7 @@ import bisqRoutes from './api/bisq/bisq.routes';
 import liquidRoutes from './api/liquid/liquid.routes';
 import bitcoinRoutes from './api/bitcoin/bitcoin.routes';
 import fundingTxFetcher from "./tasks/lightning/sync-tasks/funding-tx-fetcher";
+import opBlockHeaderService from './oe/service/op-block-header.service';
```
this line should be removed

Natsvora replied:
done

dambaev commented on backend/src/oe/api/database-migration.ts:
```diff
@@ -190,6 +194,31 @@ class OpEnergyDatabaseMigration {
+    const query = `CREATE TABLE \`blockheaders\` (
+      \`height\` int unsigned NOT NULL DEFAULT 0,
+      \`version\` int unsigned NOT NULL DEFAULT 0,
+      \`previous_block_hash\` varchar(65) DEFAULT NULL,
```
I haven't been followed topic too strictly, so for me storing hash of the previous block without storing hash of the current block is weird. Can you explain the use case?

For example, to get the hash of the last stored block, it will be needed to wait for new block to arrive, which seems not correct to me.

Other than that, it seems to me that this field is the only one, which can have DEFAULT value and be NULL. Other fields should have no DEFAULT values as for me as  well as should be NOT NULL

Natsvora replied:
we just want a copy of the already existing blocks table. so we should have both fields I think. I added the current block hash also

dambaev commented on backend/src/oe/api/database-migration.ts:
```diff
@@ -190,6 +194,31 @@ class OpEnergyDatabaseMigration {
+      \`difficulty\` double unsigned NOT NULL DEFAULT 0,
+      \`nonce\` bigint(20) unsigned NOT NULL DEFAULT 0,
+      \`reward\` bigint(20) unsigned NOT NULL DEFAULT 0,
+      \`mediantime\` bigint(20) unsigned,
```
shouldn't be NULL

Natsvora replied:
fixed

dambaev commented on backend/src/api/blocks.ts:
```diff
@@ -461,7 +461,6 @@ class Blocks {
             indexer.reindex();
           }
           await blocksRepository.$saveBlockInDatabase(blockExtended);
-
```
this file belongs to mempool's code base, so I suggest to revert this diff as it can give a rebase conflict in the future

Natsvora replied:
Done

dambaev commented on backend/src/oe/api/database-migration.ts:
```diff
@@ -190,6 +194,31 @@ class OpEnergyDatabaseMigration {
+      const err_msg = `OE MIGRATION: $createTableBlockHeaders error ${( e instanceof Error ? e.message : e)}`;
+      throw new Error( err_msg);
+    }
+    logger.info( 'OE MOGRATION: OpEneryDatabaseMigration.$createTableBlockHeaders completed');
```
I have noticed, that I have a mistake in word MIGRATION ;)

Natsvora replied:
Done fixed at all the places

dambaev commented on backend/src/oe/database.ts:
```diff
@@ -52,6 +52,44 @@ export class DB {
+  private static async $blockSpanPool_getConnection(UUID: string): Promise<PrivatePoolConnection> {
+    logger.info( `${UUID} PROFILE: start: blockSpanPool.getConnection`);
+    const connection = await private_DB.blockSpanPool.getConnection();
+    private_DB.connections_count++;
```
this field was supposed to track connections count of the accountPool. If we will have one more pool (blockSpanPool), then there should be distinct variable to track connections count for appropriate pool.

The goal is to be able to handle the case when we will reach connections limit from the logs.

Natsvora replied:
Added new variable to track blockSpan connection count

dambaev commented on backend/src/oe/database.ts:
```diff
@@ -52,6 +52,44 @@ export class DB {
+    return ({ value: connection} as PrivatePoolConnection);
+  }
+  
+  public static async $blockSpanPool_query<T extends RowDataPacket[][] | RowDataPacket[] | OkPacket | OkPacket[] | ResultSetHeader>( UUID: string, connection: PrivatePoolConnection, query: string, args: any | any[] | { [param: string]: any }): Promise<[T, FieldPacket[]]> {
```
for me it seems, that `$blockSpanPool_query` differs from `$accountPool_query` only with log messages. I suggest to keep only 1 of them with renaming it into `$profile_query` like this:
```
public static async $profile_query<T extends RowDataPacket[][] | RowDataPacket[] | OkPacket | OkPacket[] | ResultSetHeader>( UUID: string, label: string, connection: PrivatePoolConnection, args: any | any[] | { [param: string]: any }, query: string = ''): Promise<[T, FieldPacket[]]>
{
  logger.info( `${UUID} PROFILE query ${label}: start`);
  var result = await connection.value.query<T>( query, args);
  logger.info( `${UUID} PROFILE query ${label}: end`);
  return result;
}
```
this way we will have 1 generic function + we will be able to use `label` to provide some labels to use them as 'costs' for performance profiling reasons

Natsvora replied:
Why are we using var? instead of let and const? any specific reason for it?

dambaev replied:
you are correct here: const should be enough for this

Natsvora replied:
@dambaev I suggest adding the last optional parameter label with the default value accountPool. I did not suggest modifying accountPool_query method which is used at more than 20 places. It is fine we can rename it profile_query but did not suggest changing the arguments it may create some bugs if we did any mistake

dambaev commented on backend/src/oe/api/op-statistics.service.ts:
```diff
@@ -1,7 +1,7 @@
+import opBlockHeaderService from '../service/op-block-header.service';
+import { BlockHeader, BlockHeight } from './interfaces/op-energy.interface';
 
 export class OpStatisticService {
```
As far as I can see: there are many algorithms for calculating variance (and std dev is sqrt(variance)), so I am not sure if this one is optimal, but I can propose to introduce a table nbdrStatisticCache( blockHeight, blockSpan, avg, stddev) to not to recalculate result multiple times

Natsvora replied:
Please ignore this as I and Thomas are improving this moving forward. We have some plans to optimize this.

Commit:
[147] Updated block header logic

Commit:
[147] Added changes related to data types

Natsvora commented:
@dambaev Can you please review this again I did the changes.

dambaev commented on backend/src/oe/repositories/OpBlockHeaderRepository.ts:
```diff
@@ -0,0 +1,97 @@
+    }
+  }
+
+  public async $getBlock(UUID: string, height: number): Promise<BlockHeader> {
```
height should be of type ConfirmedBlockHeight as we know, that unconfirmed block are definitely missing from DB

Natsvora replied:
Done

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,137 @@
+import opBlockHeaderRepository from '../repositories/OpBlockHeaderRepository';
+
+export class OpBlockHeaderService {
+  async $getBlockHeaderData(blockHeight: ConfirmedBlockHeight) {
```
this function is missing return type definition

Natsvora replied:
Done

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,137 @@
+  async $getBlockHeaderData(blockHeight: ConfirmedBlockHeight) {
+    try {
+
+      // Only storing 6th block from current tip
```
this comment is not related to any code within this function, I guess it's some kind of leftover?

Natsvora replied:
Removed

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,137 @@
+    }
+  }
+
+  public async $getBlockHeadersByHeights(UUID: string, bockHeights: number[]): Promise<BlockHeader[]> {
```
is there a reason to not to use `OpBlockHeaderRepository.$getBlockHeadersByHeights` instead of `OpBlockHeaderService.$getBlockHeadersByHeights`?

Natsvora replied:
getBlockHeadersByHeights is used while calculating nbdr I prefer to use the service instead of directly calling the repository in the nbdr service. 

So nbdr service is just consuming another service OpBlockHeaderService

dambaev commented on backend/src/oe/service/op-block-header.service.ts:
```diff
@@ -0,0 +1,137 @@
+    }
+  }
+
+  public async $getBlockHeader(UUID: string, height: number): Promise<BlockHeader | null> {
```
why not to use `opBlockHeaderRepository.$getBlock` directly?

Natsvora replied:
I am not using this in my code this is for future use if some one wants to get block heder so they can fetch it.

dambaev replied:
well, if you think, that users should use service `OpBlockHeaderService.$getBlockHeader` instead of `opBlockHeaderRepository.$getBlock`, then height should be of type ConfirmedBlockHeight.

I should mention is that every time you are constructing ConfirmedBloockHeight by using `{ value: Something}` instead of using verifyConfirmedBlockHeight() function you are basically defeating the purpose of ConfirmedBlockHeight, which purpose is to make sure, that caller had confirmed that given block height is less or equal tip - 6...

So ideally, there should be no way to create ConfirmedBlockHeight value without verfiyConfirmedBlockHeight() function, but, unfortunately, TypeScript does not allow such granularity. But if you do this, then I expect that there will be some check of value, that is missing....

dambaev commented on backend/src/oe/api/op-statistics.service.ts:
```diff
@@ -46,15 +44,14 @@ export class OpStatisticService {
               .reduce((a, b) => a + b) /
-              length -
-              1
+            (length - 1)
```
```
            nbdrStatisticsList
              .map((x) => Math.pow(x - mean, 2))
              .reduce((a, b) => a + b) /
```
2 * O(n) time =>
```
            nbdrStatisticsList
              .reduce((a, x) => a + (Math.pow(x - mean, 2))) /
```
O(n) time

Natsvora replied:
Good suggestion. I missed it

dambaev commented on backend/src/oe/api/op-statistics.service.ts:
```diff
@@ -1,42 +1,40 @@
+      let lastblock = blockHeight.value;
+      const blockNumbers = [] as number[];
+      // creating array of last 100 blocks spans
+      for (let i = 0; i < 100; i += 2) {
```
```
for (let i = 0; i < 50; i ++) {
```
?

Natsvora replied:
I need even and in your case and in my case both times it will only iterate 50 times so don't make any difference

dambaev commented on backend/src/oe/api/op-statistics.service.ts:
```diff
@@ -1,42 +1,40 @@
+      const nbdrStatisticsList: number[] = [];
+      let lastblock = blockHeight.value;
+      const blockNumbers = [] as number[];
+      // creating array of last 100 blocks spans
```
no, we are creating array for 50 blockspans as blockspan is a range defined by start and end, isn't it?

Natsvora replied:
Humm ii need to see that. Please ignore this now in this PR we will discuss that in detail In nbdr task

Commit:
147 - Added confirmed blocks logic

Natsvora commented:
@dambaev Added Changes please review it again.

dambaev commented on backend/src/oe/repositories/OpBlockHeaderRepository.ts:
```diff
@@ -0,0 +1,97 @@
+    }
+  }
+
+  public async $getBlockHeadersByHeights(UUID: string, blockHeights: number[]): Promise<BlockHeader[]> {
```
I should suggest to make blockHeights of type ConfirmedBlockHeight[]. The use case, for example: it will force nbdr statistics to check heights: it will turn out, that we can't get 50 blockspans for some range of inputs...

Natsvora replied:
Done. Enforced use of the confirmed block.

Commit:
147: enforced use of confirmed block
